examples/jwt/src/main.rs

```diff
@@ -0,0 +1,178 @@
+//! Example JWT authorization/authentication.
+//!
+//! Run with
+//!
+//! ```not_rust
+//! cargo run -p example-jwt
+//! ```
+
+use axum::{
+    body::{Bytes, Full},
+    extract::TypedHeader,
+    handler::{get, post},
+    http::{Response, StatusCode},
+    response::IntoResponse,
+    Json, Router,
+};
+use headers::{authorization::Bearer, Authorization};
+use jsonwebtoken::{decode, encode, DecodingKey, EncodingKey, Header, Validation};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::{convert::Infallible, fmt::Display, net::SocketAddr};
+
+// Quick instructions
+//
+// - get an authorization token:
+//
+// curl -s \
+//     -w '\n' \
+//     -H 'Content-Type: application/json' \
+//     -d '{"client_id":"foo","client_secret":"bar"}' \
+//     http://localhost:3000/auth
+//
+// - visit the protected area using the token for authentication
+//
+// curl -s \
+//     -w '\n' \
+//     -H 'Content-Type: application/json' \
+//     -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJiQGIuY29tIiwiY29tcGFueSI6IkFDTUUiLCJleHAiOjEwMDAwMDAwMDAwfQ.M3LAZmrzUkXDC1q5mSzFAs_kJrwuKz3jOoDmjJ0G4gM' \
+//     http://localhost:3000/protected
+//
+// - try to visit the protected area using an invalid token
+//
+// curl -s \
+//     -w '\n' \
+//     -H 'Content-Type: application/json' \
+//     -H 'Authorization: Bearer blahblahblah' \
+//     http://localhost:3000/protected
+
+static JWT_SECRET: &[u8] = b"secret";
+
+#[tokio::main]
+async fn main() {
+    // Set the RUST_LOG, if it hasn't been explicitly defined
+    if std::env::var("RUST_LOG").is_err() {
+        std::env::set_var("RUST_LOG", "example_jwt=debug")
+    }
+    tracing_subscriber::fmt::init();
+
+    let app = Router::new()
+        .route("/protected", get(protected))
+        .route("/auth", post(auth));
+
+    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
+    tracing::debug!("listening on {}", addr);
+
+    axum::Server::bind(&addr)
+        .serve(app.into_make_service())
+        .await
+        .unwrap();
+}
+
+async fn protected(
+    TypedHeader(authorization): TypedHeader<Authorization<Bearer>>,
+) -> Result<String, AuthError> {
+    // Extract the authorization header value (i.e the token)
+    let token = authorization.0.token();
+
+    // Authenticate the user and returns his data
+    let validation = Validation {
+        sub: Some("b@b.com".to_string()),
+        ..Validation::default()
+    };
+    let token_data = decode::<Claims>(token, &DecodingKey::from_secret(JWT_SECRET), &validation)
+        .map_err(|_| AuthError::InvalidToken)?;
```

Comment:
I imagine this is something users are gonna do a lot. What do you think about wrapping it up into its own extractor, so its. easier to reuse for other routes? Probably good to show even though its only used in one route here.

Reply:
I totally agree. Tonight I'll read the extractors documentation to check how to implement them. Any example showing it will be appreciated. :smiley:

Reply:
Done in 6016c65.